cli/commons/src/Microsoft.Kiota.Cli.Commons/IO/ODataPagingService.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.Kiota.Abstractions;
+
+[assembly: InternalsVisibleTo("Microsoft.Kiota.Cli.Commons.Tests")]
+namespace Microsoft.Kiota.Cli.Commons.IO;
+
+/// <summary>
+/// Paging service that supports the x-ms-pageable extension
+/// </summary>
+public class ODataPagingService : IPagingService
+{
+    /// <inheritdoc />
+    public async Task<Uri?> GetNextPageLinkAsync(PageLinkData pageLinkData, CancellationToken cancellationToken = default)
+    {
+        if (pageLinkData.ResponseFormat == ResponseFormat.JSON)
+        {
+            try
+            {
+                var doc = await JsonDocument.ParseAsync(pageLinkData.Response, cancellationToken: cancellationToken);
+                if (doc.RootElement.ValueKind == JsonValueKind.Object)
+                {
+                    var obj = doc.RootElement.EnumerateObject().FirstOrDefault(o => o.Name == pageLinkData.NextLinkName);
+                    string? link = null;
+                    if (obj.Value.ValueKind == JsonValueKind.String)
+                        link = obj.Value.GetString();
+                    if (!string.IsNullOrWhiteSpace(link)) return new Uri(link);
+                }
+            }
+            catch (JsonException)
+            {
+                // If the response isn't valid JSON, there will be no next link.
+                // TODO: Log warning once logging story is defined
+            }
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public async Task<Stream> GetPagedDataAsync(Func<RequestInformation, CancellationToken, Task<Stream>> requestExecutorAsync, PageLinkData pageLinkData, bool fetchAllPages = false, CancellationToken cancellationToken = default)
+    {
+        // Set the page size to 999 if the user asked to fetch all pages and top either isn't specified or is invalid
+        if (fetchAllPages && (!pageLinkData.RequestInformation.QueryParameters.TryGetValue("%24top", out var topVal) || (topVal as int?) < 1))
+        {
+            pageLinkData.RequestInformation.QueryParameters["%24top"] = 999;
```

Comment:
The service is OData specific. That's why I've used top. Other APIs will implement the IPagingService interface differently

Reply:
unless I'm missing something, setting top is the only reason why this PagingService would need to be specialized for OData, correct?
I'm a bit reluctant to add anything specific to OData in the core of Kiota, because it breaks one of our promises.
What would you think of having a base paging service it, protocol agnostic, which offers a callback to specialize the request information, and then have a derived ODataPagingService in Graph core which implements this specific behaviour instead?

Reply:
There was a request to have sensible defaults for features of the Kiota generated CLIs. If someone were to generate a CLI, being able to get up and running without having to implement interfaces first provides a great getting started experience. With paging, I thought that default would be OData paging that Graph uses.

Reply:
Do you have a suggestion for what we can use as a default? I could also remove that section as it isn't really necessary for the logic to work. It's more of a sanitation on the input parameters.

Reply:
default query parameter name?

Reply:
Default paging protocol

Reply:
unresolving as this is still part of the code for now